### PR TITLE
Add support for mutable access

### PR DESCRIPTION
**Pull request by hawkw**
This is still quite rough, needs docs, additional testing, and cleanup, but it basically works.

Signed-off-by: Eliza Weisman <eliza@buoyant.io>

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -566,6 +609,95 @@ where
...
+    }
+}
+
+impl std::error::Error for GetMutError {}
+
+impl GetMutError {
+    pub(crate) const BORROWED: Self = Self { kind: GetMutErrorKind::InUse, };
+    pub(crate) const NONEXISTENT: Self = Self { kind: GetMutErrorKind::Nonexistent, };
+
+    pub fn is_borrowed(&self) -> bool {
```
Comment (bIgBV) [inline src/lib.rs:686]: are there going to be more than two states? An if expression would be easier to read here imo.

```suggestion
    pub fn is_borrowed(&self) -> bool {
        self.kind == GetMutErrorKind::InUse
    }
```

File: src/lib.rs [hunk 2]
```diff
@@ -566,6 +609,95 @@ where
...
+        match self.kind {
+            GetMutErrorKind::InUse => true,
+            _ => false,
+        }
+    }
+
+    pub fn is_nonexistent(&self) -> bool {
```
Comment (bIgBV) [inline src/lib.rs:693]: Same here as above.

**Comment by bIgBV** [comment]
Having gone through the changes I think you got the main parts all done. I've got one thought though. I think we can provide more granular error messages here. Now we have multiple failure conditions for why a `get` call fails:

* Shard is full
* Someone else has already removed the value present in that slot (generation has updated)
* Someone else is has exclusive access to it. 

As a user, I'd like to know which of these three cases caused the failure. I see that you already created a user facing error type `GetMutError` with a couple of errors, but I think we need to change methods on `Slot` to return a `Result<T, SlotError>` instead of an `Option`.

**Comment by bIgBV** [comment]
Besides that, I think it's mostly just tests and docs.

**Comment by bIgBV** [comment]
@hawkw Added tests for the slab impl and looks like you had already created tests for the pool impl. Also added documentation. I figured I could port the `create` API to return a `get_mut` guard in a separate PR.

File: src/lib.rs [hunk 3]
```diff
@@ -566,6 +634,99 @@ where
...
+        match self.kind {
+            GetMutErrorKind::Nonexistent => true,
+            _ => false,
+        }
```
Comment (hawkw) [inline src/lib.rs:726]:
```suggestion
        self.kind == GetMutErrorKind::Nonexistent
```

File: src/lib.rs (hunk 1 above)
Comment (hawkw) [reply src/lib.rs:715]: Good call, I forgot that the kind enum was `Eq`.

**Pull request closed without merging**